### DISCUSSION TITLE
Misc Question: How does this compare to tsup?

### DISCUSSION THREAD

**Discussion by jessekrubin** [privatenumber/pkgroll#18] [Q&A] [answered]
Thanks!!!

**Comment by privatenumber** [comment, marked as answer]
They are similar bundlers, but I think the main differences are:
- _pkgroll_ is zero-config. It reads the entry-points declared in your `package.json` to determine how to bundle the package. _tsup_ requires manual configuration.

- _pkgroll_ is a thin abstraction over Rollup (just smartly configures it). Similarly, _tsup_ is a thin abstraction over esbuild. _pkgroll_ also uses esbuild for transformations & minification as a Rollup plugin, but the bundling & tree-shaking is done by Rollup (which is [known to output best/cleanest code](https://github.com/privatenumber/pkgroll#why-bundle-with-rollup)). However, when _tsup_ emits type declaration, it also uses Rollup which eliminates the speed gains from using esbuild.

- IIRC because _tsup_ uses esbuild, the ESM to CJS compilation wasn't great compared to Rollups. As a package maintainer who wants to support both ESM and CJS exports, this was one of the biggest annoyances of using _tsup_ (which compelled me to develop _pkgroll_).

**Reply by jessekrubin** [reply]
Thanks so much for your reply!
